src/network/networkd-route.c

```diff
@@ -140,8 +140,10 @@ void route_free(Route *route) {
         network_config_section_free(route->section);
 
         if (route->link) {
-                set_remove(route->link->routes, route);
-                set_remove(route->link->routes_foreign, route);
+                if (set_get(route->link->routes, route) == route)
+                        set_remove(route->link->routes, route);
```

Comment:
Sorry I am not able to follow this patch if there is no route it will not be removed so why we checking ?

Reply:
Let's consider the case that there exists two `Route` object `a` and `b` such that they are equivalent in the base of `route_hash_func()`, and `a` is stored in the `Set`s. When `b` is freed, then the entry of `a` is removed from the `Set`s, thus `a` is leaked.

Of course, this is a 'potential' memory leak at this time, and the above path did not occurs. This is just a preparation for the following commits.

Reply:
@keszybz and @ssahani After re-considering the code, I conclude that this change is not necessary. Hence, dropped now.